tests/tools/Test_CompileReleaseNotes.py

```diff
@@ -0,0 +1,156 @@
+            "as before.")
+        self.assertIsNone(
+            get_upgrade_instructions(
+                "<!-- UPGRADE INSTRUCTIONS --> <!-- UPGRADE INSTRUCTIONS -->"),
+            "Not all whitespace is stripped")
+        self.assertIsNone(
+            get_upgrade_instructions(
+                "<!-- UPGRADE INSTRUCTIONS -->\n<!-- UPGRADE INSTRUCTIONS -->"
+            ), "Not all whitespace is stripped")
+        self.assertEqual(
+            get_upgrade_instructions(
+                ("<!-- UPGRADE INSTRUCTIONS -->\n\n\n   Do this and that."
+                 "  \n\n<!-- UPGRADE INSTRUCTIONS -->")), "Do this and that.",
+            "Not all whitespace is stripped")
+
+    def test_compile_release_notes(self):
+        self.assertEqual(
+            compile_release_notes([]),
+            textwrap.dedent("""\
+                ## Merged pull-requests
+
+                _None_
+                """))
+        pr1 = PullRequest(id=1, title="Add this")
+        pr2 = PullRequest(id=2,
+                          title="Also add this new feature",
+                          url="https://github.com/2")
+        self.assertEqual(
+            compile_release_notes([pr1, pr2]),
+            textwrap.dedent("""\
+                ## Merged pull-requests
+
+                **General changes:**
+
+                - Add this (#1)
+                - Also add this new feature ([#2](https://github.com/2))
+                """))
+        major_pr1 = PullRequest(id=3,
+                                title="This is big",
+                                group='major new feature',
+                                upgrade_instructions="- Do this.\n- And that.")
+        major_pr2 = PullRequest(id=4,
+                                title="Another feature",
+                                group='major new feature')
+        bugfix_pr1 = PullRequest(
+            id=5,
+            title="Fixed this bug",
+            url='https://github.com/5',
+            group='bugfix',
+            upgrade_instructions="You'll have to rerun your simulation.")
+        bugfix_pr2 = PullRequest(id=6,
+                                 title="Fixed another bug",
+                                 group='bugfix')
+        self.assertEqual(
+            compile_release_notes(
+                [pr1, major_pr1, pr2, bugfix_pr1, bugfix_pr2, major_pr2]),
+            textwrap.dedent("""\
+                ## Upgrade instructions
+
+                **From #3 (This is big):**
+
+                - Do this.
+                - And that.
+
+                **From [#5](https://github.com/5) (Fixed this bug):**
+
+                You'll have to rerun your simulation.
+
+                ## Merged pull-requests
+
+                **Major new features:**
+
+                - This is big (#3)
+                - Another feature (#4)
+
+                **General changes:**
+
+                - Add this (#1)
+                - Also add this new feature ([#2](https://github.com/2))
```

Comment:
Why are these linked sometimes, but not all of the time?

Reply:
Because I added the `url` field to only some of the test-objects. If no `url` is available then the ID is not linked. In practice, URLs will be available when GitHub is queried, and won't be available when GitHub-queries are disabled with the `--no-github` flag.